pkg/project/auth/cache.go

```diff
@@ -0,0 +1,359 @@
+func (ac *AuthorizationCache) addSubjectsToNamespace(subjectRecordStore cache.Store, subjects []string, namespace string) {
+	for _, subject := range subjects {
+		var item *subjectRecord
+		obj, exists, _ := subjectRecordStore.GetByKey(subject)
+		if exists {
+			item = obj.(*subjectRecord)
+		} else {
+			item = &subjectRecord{subject: subject, namespaces: util.NewStringSet()}
+			subjectRecordStore.Add(item)
+		}
+		item.namespaces.Insert(namespace)
+	}
+}
+
+// cacheReviewRecord updates the cache based on the request processed
+func (ac *AuthorizationCache) cacheReviewRecord(request *reviewRequest, lastKnown *reviewRecord, review Review, reviewRecordStore cache.Store) {
+	reviewRecord := &reviewRecord{
+		reviewRequest: &reviewRequest{namespace: request.namespace, policyUIDToResourceVersion: map[types.UID]string{}, policyBindingUIDToResourceVersion: map[types.UID]string{}},
+		groups:        review.Groups(),
+		users:         review.Users(),
+	}
+	// keep what we last believe we knew by default
+	if lastKnown != nil {
+		reviewRecord.namespaceResourceVersion = lastKnown.namespaceResourceVersion
+		for k, v := range lastKnown.policyUIDToResourceVersion {
+			reviewRecord.policyUIDToResourceVersion[k] = v
+		}
+		for k, v := range lastKnown.policyBindingUIDToResourceVersion {
+			reviewRecord.policyBindingUIDToResourceVersion[k] = v
+		}
+	}
+
+	// update the review record relative to what drove this request
+	if len(request.namespaceResourceVersion) > 0 {
+		reviewRecord.namespaceResourceVersion = request.namespaceResourceVersion
+	}
+	for k, v := range request.policyUIDToResourceVersion {
+		reviewRecord.policyUIDToResourceVersion[k] = v
+	}
+	for k, v := range request.policyBindingUIDToResourceVersion {
+		reviewRecord.policyBindingUIDToResourceVersion[k] = v
+	}
+	// update the cache record
+	reviewRecordStore.Add(reviewRecord)
+}
+
+func (ac *AuthorizationCache) lastKnown(reviewRecordStore cache.Store, namespace string) (*reviewRecord, error) {
+	obj, exists, err := reviewRecordStore.GetByKey(namespace)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return obj.(*reviewRecord), nil
+	}
+	return nil, nil
+}
+
+// syncRequest takes a reviewRequest and determines if it should update the caches supplied, it is not thread-safe
+func (ac *AuthorizationCache) syncRequest(request *reviewRequest, userSubjectRecordStore cache.Store, groupSubjectRecordStore cache.Store, reviewRecordStore cache.Store) error {
+
+	lastKnown, err := ac.lastKnown(reviewRecordStore, request.namespace)
+	if err != nil {
+		return err
+	}
+
+	if ac.skipReview(request, lastKnown) {
+		return nil
+	}
+
+	namespace := request.namespace
+	review, err := ac.reviewer.Review(namespace)
+	if err != nil {
+		return err
+	}
+	users := util.NewStringSet(review.Users()...)
+	groups := util.NewStringSet(review.Groups()...)
+
+	// TODO: switch to use the diff function I pushed upstream
+	usersToRemove := util.NewStringSet()
```

Comment:
To remove is simply `Insert(oldUsers.List()).Delete(newUsers.List())`  The only ones left in the set are those were in old and not in new.